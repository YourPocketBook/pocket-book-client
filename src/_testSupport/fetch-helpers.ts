import fetchMock, {
  MockOptionsMethodDelete,
  MockOptionsMethodGet,
  MockOptionsMethodPost,
  MockOptionsMethodPut,
  MockResponse,
  MockResponseFunction
} from "fetch-mock";

afterEach(fetchMock.reset);

export function authenticatedGet(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodGet
) {
  fetchMock.get(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}

export function authenticatedPut(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPut
) {
  fetchMock.put(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedPost(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPost
) {
  fetchMock.post(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedDelete(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodDelete
) {
  fetchMock.delete(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}
