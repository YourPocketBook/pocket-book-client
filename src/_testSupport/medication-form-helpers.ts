import { fireEvent } from "@testing-library/react";

export const testDefaultState = {
  adviceIfDeclined: "Test Advice 1",
  adviceIfTaken: "Test Advice 2",
  dose: "Test Dose",
  exclusionCriteria: "Test Exclusion",
  form: "Test Form",
  inclusionCriteria: "Test Inclusion",
  indications: "Test Indications",
  name: "Test Name",
  policyDate: "2019-12-10",
  route: "Test Route",
  sideEffects: "Test Side Effects"
};

export function setupState(getByLabelText: (text: string) => HTMLElement) {
  fireEvent.change(getByLabelText("Name"), {
    target: { value: testDefaultState.name }
  });
  fireEvent.change(getByLabelText("Policy Date"), {
    target: { value: testDefaultState.policyDate }
  });
  fireEvent.change(getByLabelText("Indications"), {
    target: { value: testDefaultState.indications }
  });
  fireEvent.change(getByLabelText("Inclusion Criteria"), {
    target: { value: testDefaultState.inclusionCriteria }
  });
  fireEvent.change(getByLabelText("Exclusion Criteria"), {
    target: { value: testDefaultState.exclusionCriteria }
  });
  fireEvent.change(getByLabelText("Name, Forms and Strengths of Medicine"), {
    target: { value: testDefaultState.form }
  });
  fireEvent.change(getByLabelText("Route or Method"), {
    target: { value: testDefaultState.route }
  });
  fireEvent.change(getByLabelText("Dosage"), {
    target: { value: testDefaultState.dose }
  });
  fireEvent.change(getByLabelText("Side Effects"), {
    target: { value: testDefaultState.sideEffects }
  });
  fireEvent.change(getByLabelText("Advice if Taken"), {
    target: { value: testDefaultState.adviceIfTaken }
  });
  fireEvent.change(getByLabelText("Advice if Declined"), {
    target: { value: testDefaultState.adviceIfDeclined }
  });
}

export function checkForClass(
  getByLabelText: (text: string) => HTMLElement,
  cls: "is-valid" | "is-invalid"
) {
  expect(getByLabelText("Name")).toHaveClass(cls);
  expect(getByLabelText("Policy Date")).toHaveClass(cls);
  expect(getByLabelText("Indications")).toHaveClass(cls);
  expect(getByLabelText("Inclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Exclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Name, Forms and Strengths of Medicine")).toHaveClass(
    cls
  );
  expect(getByLabelText("Route or Method")).toHaveClass(cls);
  expect(getByLabelText("Dosage")).toHaveClass(cls);
  expect(getByLabelText("Side Effects")).toHaveClass(cls);
  expect(getByLabelText("Advice if Taken")).toHaveClass(cls);
  expect(getByLabelText("Advice if Declined")).toHaveClass(cls);
}

export function checkState(
  getByLabelText: (text: string) => HTMLElement,
  hasValue: boolean
) {
  expect(getByLabelText("Name")).toHaveValue(
    hasValue ? testDefaultState.name : ""
  );
  expect(getByLabelText("Policy Date")).toHaveValue(
    hasValue ? testDefaultState.policyDate : ""
  );
  expect(getByLabelText("Indications")).toHaveTextContent(
    hasValue ? testDefaultState.indications : ""
  );
  expect(getByLabelText("Inclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.inclusionCriteria : ""
  );
  expect(getByLabelText("Exclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.exclusionCriteria : ""
  );
  expect(
    getByLabelText("Name, Forms and Strengths of Medicine")
  ).toHaveTextContent(hasValue ? testDefaultState.form : "");
  expect(getByLabelText("Route or Method")).toHaveTextContent(
    hasValue ? testDefaultState.route : ""
  );
  expect(getByLabelText("Dosage")).toHaveTextContent(
    hasValue ? testDefaultState.dose : ""
  );
  expect(getByLabelText("Side Effects")).toHaveTextContent(
    hasValue ? testDefaultState.sideEffects : ""
  );
  expect(getByLabelText("Advice if Taken")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfTaken : ""
  );
  expect(getByLabelText("Advice if Declined")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfDeclined : ""
  );
}
