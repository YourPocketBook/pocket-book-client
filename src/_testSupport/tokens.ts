export const inDateNonAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.VmT5YkTG0sUd5sK7skfMdi9kc5XG7Yhx6sqhqMSxhVY";
export const inDateAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInJvbGUiOiJBZG1pbiIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.rSQb2Hv6RfxaXSfO6s3QjLjnYKeqVgps3rPenquh3Nk";
export const expiredToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE0MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInVuaXF1ZV9uYW1lIjoiVGVzdCBVc2VyIiwibmFtZWlkIjoiZmYxZDA4ZTctNDY1Ni00YzI4LTk3NmYtMWMyN2YxOTE0NzdmIiwiZW1haWwiOiJ0ZXN0QHNqYS5vcmcudWsifQ.7bDfHDk9sWXL8boFeZWIdo_GB-4Qj8PvA4HDC49EzNc";
