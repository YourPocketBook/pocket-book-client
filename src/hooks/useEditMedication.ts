import isValid from "date-fns/isValid";
import parseISO from "date-fns/parseISO";
import { useReducer } from "react";
import {
  IEditMedication,
  IEditMedicationValid
} from "../medications/medication-form";

export function useEditMedication() {
  const [state, dispatch] = useReducer(
    (
      s: IEditMedication & { valid: IEditMedicationValid },
      a: { prop: keyof IEditMedication } & (
        | { op: "update"; value: string | undefined | null }
        | {
            op: "invalidate";
          }
        | { op: "setValid"; value: boolean }
      )
    ) => {
      if (s.hasOwnProperty(a.prop)) {
        const newData = { ...s };

        switch (a.op) {
          case "update":
            newData[a.prop] = a.value || "";
            if (a.prop === "policyDate") {
              newData.valid[a.prop] = !!a.value && isValid(parseISO(a.value));
            } else {
              newData.valid[a.prop] = !!a.value && a.value.length > 0;
            }
            break;
          case "invalidate":
            newData.valid[a.prop] = false;
            break;
          case "setValid":
            newData.valid[a.prop] = a.value;
            break;
        }
        return newData;
      }
      return s;
    },
    {
      adviceIfDeclined: "",
      adviceIfTaken: "",
      dose: "",
      exclusionCriteria: "",
      form: "",
      inclusionCriteria: "",
      indications: "",
      name: "",
      policyDate: "",
      route: "",
      sideEffects: "",
      valid: {
        adviceIfDeclined: false,
        adviceIfTaken: false,
        dose: false,
        exclusionCriteria: false,
        form: false,
        inclusionCriteria: false,
        indications: false,
        name: false,
        policyDate: false,
        route: false,
        sideEffects: false
      }
    }
  );

  return { state, dispatch };
}
