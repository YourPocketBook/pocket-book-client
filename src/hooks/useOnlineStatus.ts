// Included from https://github.com/21kb/react-hooks/blob/master/packages/react-online-status-hook
//
// Copyright Â© 2018 Kunall Banerjee
// MIT License

import { useEffect, useState } from "react";

export function useOnlineStatus() {
  const [state, setState] = useState(navigator.onLine);

  const onOnlineEvent = () => {
    setState(navigator.onLine);
  };

  const onOfflineEvent = () => {
    setState(navigator.onLine);
  };

  useEffect(() => {
    window.addEventListener("online", onOnlineEvent);
    window.addEventListener("offline", onOfflineEvent);

    return () => {
      window.removeEventListener("online", onOnlineEvent);
      window.removeEventListener("offline", onOfflineEvent);
    };
  });

  return state;
}
