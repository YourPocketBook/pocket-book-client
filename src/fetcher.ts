import { getToken, login, userAgentApplication } from "./ms-login";
import { toDate, isFuture } from "date-fns";

class Fetcher {
  public async authFetch(
    input: RequestInfo,
    init?: RequestInit
  ): Promise<Response> {
    const token = await getToken();

    if (!token) {
      login();
      return Promise.reject("No token");
    }

    init = {
      ...init,
      headers: {
        ...(init ? init.headers : {}),
        Authorization: `Bearer ${token}`
      }
    };

    return fetch(input, init);
  }

  public getName() {
    const userAccount = userAgentApplication.getAccount();

    if (!userAccount) {
      return "";
    }
    return userAccount.name;
  }

  public hasToken(): boolean {
    return !!userAgentApplication.getAccount();
  }

  public isInDate() {
    const userAccount = userAgentApplication.getAccount();

    return (
      !!userAccount &&
      isFuture(toDate(parseInt(userAccount.idTokenClaims.exp, 10) * 1000))
    );
  }
}

const fetcher = new Fetcher();

export { fetcher, Fetcher };
