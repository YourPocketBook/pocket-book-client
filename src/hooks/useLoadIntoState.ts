import { Dispatch, useEffect, useState } from "react";
import { fetcher } from "../fetcher";

/**
 * Fetches JSON from the specified uri and loads each property into individually editable state variables
 *
 * @param uri The URI to load the data from
 * @param setState An object containing setState functions for each property
 * @param [authenticate=false] If set to true, the current token will be passed with the call
 * @returns True if the request is currently running, otherwise false.
 */
export function useLoadIntoEditableState<
  TData extends { [key: string]: string | undefined | null } & object
>(
  uri: string,
  dispatch: Dispatch<{
    op: "update";
    prop: keyof TData;
    value: string | undefined | null;
  }>,
  authenticate = false
): boolean {
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const fetchAbort = new AbortController();

    const options: RequestInit = {
      cache: "reload",
      method: "GET",
      signal: fetchAbort.signal
    };

    async function func() {
      try {
        const res = authenticate
          ? await fetcher.authFetch(uri, options)
          : await fetch(uri, options);

        if (res.ok) {
          const data = (await res.json()) as TData;
          for (const prop in data) {
            if (data.hasOwnProperty(prop)) {
              dispatch({ op: "update", prop, value: data[prop] });
            }
          }
        }
      } catch (ex) {
        if (!fetchAbort.signal.aborted) {
          // tslint:disable-next-line:no-console
          console.log(`Could not get from ${uri}.`);
        }
      } finally {
        if (!fetchAbort.signal.aborted) {
          setLoading(false);
        }
      }
    }

    func();
    return () => fetchAbort.abort();
  }, [uri, authenticate, dispatch]);

  return loading;
}
