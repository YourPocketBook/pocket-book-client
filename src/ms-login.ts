import {
  AuthError,
  AuthResponse,
  Configuration,
  UserAgentApplication
} from "msal";
import { fetcher } from "./fetcher";
import { history } from "./history";

const sjaAuthority = "91d037fb-4714-4fe8-b084-68c083b8193f";

const config: Configuration = {
  auth: {
    authority: `https://login.microsoftonline.com/${sjaAuthority}`,
    clientId: "6a5d659f-f29d-4f37-9a72-0378ae55bd6b",
    navigateToLoginRequestUrl: false
  },
  cache: {
    cacheLocation: "localStorage"
  }
};

const userAgentApplication = new UserAgentApplication(config);
userAgentApplication.handleRedirectCallback(authCallback);

function authCallback(authErr: AuthError, response?: AuthResponse) {
  if (!authErr) {
    if (
      response?.tokenType === "id_token" &&
      userAgentApplication.getAccount() &&
      response.tenantId === sjaAuthority
    ) {
      fetcher.saveToken(response.idToken.rawIdToken);
      history.push("/home");
      return;
    }
  }

  fetcher.clearToken();
  history.push("/login-error");
}

export function login() {
  const loginRequest = {
    scopes: ["https://graph.microsoft.com/User.Read"]
  };

  userAgentApplication.loginRedirect(loginRequest);
}

export function logout() {
  userAgentApplication.logout();
}
