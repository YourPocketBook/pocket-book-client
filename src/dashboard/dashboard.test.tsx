import { fireEvent, render } from "@testing-library/react";
import React from "react";
import ReactDOM from "react-dom";
import { Dashboard } from ".";
import { inDateNonAdminToken } from "../_testSupport/tokens";
import { fetcher } from "../fetcher";
import { history } from "../history";
import { useOnlineStatus } from "../hooks/useOnlineStatus";
import { logout } from "../ms-login";

jest.mock("../hooks/useOnlineStatus");
jest.mock("../ms-login");

describe("Dashboard Page", () => {
  beforeEach(() => {
    history.push = jest.fn();
    fetcher.saveToken(inDateNonAdminToken);
    (useOnlineStatus as jest.Mock).mockReturnValue(true);
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");

    ReactDOM.render(<Dashboard />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders correctly when online", () => {
    const { container } = render(<Dashboard />);

    expect(container).toMatchSnapshot();
  });

  it("renders correctly when offline", () => {
    (useOnlineStatus as jest.Mock).mockReturnValue(false);

    const { container } = render(<Dashboard />);

    expect(container).toMatchSnapshot();
  });

  it("triggers the logout process if sign out is clicked", () => {
    const { getByText } = render(<Dashboard />);

    fireEvent.click(getByText("Sign Out"));

    expect(logout).toBeCalled();
  });
});
