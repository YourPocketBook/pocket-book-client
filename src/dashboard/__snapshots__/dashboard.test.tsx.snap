// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Page renders correctly when offline 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="container"
      >
        <div
          class="alert alert-dark fade"
          role="alert"
        >
          You appear to be offline.
        </div>
        <div
          class="row"
        >
          <div
            class="col-lg-6 offset-lg-3"
          >
            <div
              class="dashboardBackground"
            >
              <div
                class="container"
              >
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <h1
                      class="text-center"
                    >
                      PocketBook
                    </h1>
                  </div>
                </div>
                <div
                  class="dashboardButton row"
                >
                  <div
                    class="col"
                  >
                    <a
                      class="btn btn-dark-green btn-block btn-lg"
                      href="/medications"
                    >
                      Medications
                    </a>
                  </div>
                </div>
                <div
                  class="dashboardButton row"
                >
                  <div
                    class="col"
                  >
                    <button
                      class="sign-out btn btn-dark-green btn-lg btn-block"
                      type="button"
                    >
                      Sign Out
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Dashboard Page renders correctly when online 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-lg-6 offset-lg-3"
          >
            <div
              class="dashboardBackground"
            >
              <div
                class="container"
              >
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <h1
                      class="text-center"
                    >
                      PocketBook
                    </h1>
                  </div>
                </div>
                <div
                  class="dashboardButton row"
                >
                  <div
                    class="col"
                  >
                    <a
                      class="btn btn-dark-green btn-block btn-lg"
                      href="/medications"
                    >
                      Medications
                    </a>
                  </div>
                </div>
                <div
                  class="dashboardButton row"
                >
                  <div
                    class="col"
                  >
                    <button
                      class="sign-out btn btn-dark-green btn-lg btn-block"
                      type="button"
                    >
                      Sign Out
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
